### DISCUSSION TITLE
list of pairs (even length)

### DISCUSSION THREAD

**Discussion by bluescreen303** [cue-lang/cue#1490] [Q&A] [answered]
I would like to define "lists with even length" and managed to do so when embedded in a struct, but I would like to be able to do this as a top level definition as well.

This works:
```
#Spec: {
  _even_length: rem(len(items), 2) & 0
  items: [...int]
}

test1: #Spec
test1: {
  items: [1,2,2,3,3,4,5,6,9,11]
}
```

cue will prevent me from creating structs with an `items` element with an odd length. Yay!

But then I tried moving this to the top level (outside of a struct):
```
#PairsList: {
  _even_length: rem(len(self), 2) & 0
  [...int]
}

test2a: #PairsList
test2a: [1,2,3]

test2b: #PairsList
test2b: [1,2,3,4]
```

But I'm missing something like a "self reference" to express this. How can this be done?

Bonus question: I'm using hidden fields in this case to express the even_length constraint. I noticed that I can also use `#even_length` which works equally well. Is there a difference between using hidden fields and # definitions?

**Comment by xinau** [comment, marked as answer]
@bluescreen303 I've stumbled up on this issue as well in my past. The solution to your problem is to [alias](https://cuelang.org/docs/references/spec/#aliases) the value. 

```cue
#PairsList: self={
  _even_length: rem(len(self), 2) & 0
  [...int]
}

test2a: #PairsList
test2a: [1,2,3]

test2b: #PairsList
test2b: [1,2,3,4]
```

There has also a discussion about adding `.` which refers to the current scope.
https://github.com/cue-lang/cue/issues/564#issuecomment-873388085 

The difference between hidden fields and definitions is better to be explained by the language reference
https://cuelang.org/docs/references/spec/#definitions-and-hidden-fields

**Reply by myitcv** [reply]
See also https://github.com/cue-lang/cue/issues/575 for a proposal regarding the `must()` builtin which would improve readability and error messages in this case.

**Reply by bluescreen303** [reply]
Thanks, `alias` was what I needed indeed.

**Reply by myitcv** [reply]
See also my comment in reply to https://github.com/cue-lang/cue/issues/1492
